protocol/packages/dex/src/impl_/transfer_out.rs

```diff
@@ -301,18 +306,18 @@ where
 
     type Result = IterNext;
 
-    type Error = Never;
+    type Error = TooManyCoins;
 
     fn visit<G>(&mut self, _coin: &CoinDTO<G>) -> StdResult<Self::Result, Self::Error>
     where
         G: Group + MemberOf<Self::GIn>,
     {
-        let next_idx = self.last_index.map_or(CoinsNb::ZERO, |prev_idx| {
-            prev_idx
-                .checked_add(1)
-                .expect("The swap task exceeds the max number of coins `CoinsNb::MAX`")
+        self.last_index = Some(if let Some(last_index) = self.last_index {
+            last_index.checked_add(1).ok_or(const { TooManyCoins })?
+        } else {
+            CoinsNb::ZERO
```

Comment:
could you revert the code to using `Option::map_or` - less branching, more concise code